torch/csrc/jit/codegen/cuda/scheduler_registry.cpp

```diff
@@ -0,0 +1,320 @@
+    //  grid reduction. This is an overkill might want to consider
+    //  trying to get the heuristics and check only if grid reduction is
+    //  required.
+    auto uses = DependencyCheck::getAllUseChains(red_tv);
+    for (auto& chain : uses) {
+      for (auto val : chain) {
+        if (val->definition()->isA<BroadcastOp>()) {
+          return false;
+        }
+      }
+    }
+
+    return true;
+  }
+
+  void schedule(Fusion* fusion) override {
+    // TODO find outputs of tv: what would we need to fill in?
+    auto red_tv = findReductionTV(fusion);
+    auto output_tv = findOutputsOfRed(fusion, red_tv);
+    scheduleReduction(fusion, rparams_, red_tv, output_tv);
+  }
+
+ private:
+  void getHeuristics(Fusion* fusion, ExpressionEvaluator& ee) {
+    auto red_tv = findReductionTV(fusion);
+    auto param = getReductionHeuristics(fusion, ee, red_tv);
+    TORCH_INTERNAL_ASSERT(param.has_value());
+    rparams_ = param.value();
+  }
+
+  TensorView* findReductionTV(Fusion* fusion) {
+    for (auto expr : fusion->exprs()) {
+      if (auto red = dynamic_cast<ReductionOp*>(expr)) {
+        if (!isTrivialReduction(red)) {
+          return red->out()->as<TensorView>();
+        }
+      }
+    }
+    TORCH_INTERNAL_ASSERT(false, "unreachable");
+    return nullptr;
+  }
+};
+
+class PointWiseScheduler : public SchedulerEntry {
+ public:
+  explicit PointWiseScheduler(Fusion* fusion)
+      : SchedulerEntry(ScheduleHeuristic::PointWise, false) {}
+
+  static bool canSchedule(Fusion* fusion) {
+    auto red_ops = findReductionOps(fusion);
+    return red_ops.empty();
+  }
+
+  void schedule(Fusion* fusion) override {
+    scheduleFusion(fusion);
+  }
+};
+
+namespace {
+
+// duplicated from Benchmark/utils.h
+static void analyzeFusion(
+    Fusion* fusion,
+    std::vector<TensorView*>& reduction_tv,
+    std::vector<TensorView*>& other_tv) {
+  auto all_values = DependencyCheck::getAllValsBetween(
+      {fusion->inputs().begin(), fusion->inputs().end()}, fusion->outputs());
+
+  for (auto tv : ir_utils::filterByType<TensorView>(all_values)) {
+    if (tv->hasReduction() && !fusion->hasInput(tv)) {
+      reduction_tv.push_back(tv);
+    } else if (!fusion->hasInput(tv)) {
+      other_tv.push_back(tv);
+    }
+  }
+}
+
+} // namespace
+class NormalizationScheduler : public SchedulerEntry {
+ public:
+  explicit NormalizationScheduler(Fusion* fusion, ExpressionEvaluator& ee)
+      : SchedulerEntry(ScheduleHeuristic::Normalization, true) {
+    getHeuristics(fusion, ee);
+  }
+
+  void schedule(Fusion* fusion) override {
+    std::vector<TensorView*> reduction_tensors;
+    std::vector<TensorView*> other_tensors;
+    analyzeFusion(fusion, reduction_tensors, other_tensors);
+    scheduleNormalization(fusion, rparams_, reduction_tensors, other_tensors);
+  }
+
+  static bool canSchedule(Fusion* fusion) {
+    auto red_ops = findReductionOps(fusion);
+
+    if (red_ops.size() < 2) {
```

Comment:
I believe the Reduction Scheduler cannot handle [Point-wise operation with Broadcast Reduction TV], since it want to inline all point-wise operations after reduction.

Reply:
I see. Thanks for pointing this out 👍 . I updated the above line to enable normalization scheduler for any reduction op count >=1. Reduction scheduler would not try to schedule a broadcast after reduction so `_softmax_backward_data` will now be handled by normalization.